### PR TITLE
-Znext-solver: adapt overflow rules to avoid breakage

### LINKED ISSUE #70: `-Ztrait-solver=next-coherence`: issue-90662-projection-caching.rs breaks
minimized https://rust.godbolt.org/z/MK6hWKhdb
```rust
trait HasProvider<T: ?Sized> {}
trait Provider<M: ?Sized> {
    type Interface: ?Sized;
}

struct Service;
struct ServiceImpl;
impl<M: HasProvider<&'static ()> + ?Sized> Provider<M> for ServiceImpl {
    type Interface = Service;
}

struct TestModule;
impl HasProvider<&'static ()> for TestModule {}
impl HasProvider<<ServiceImpl as Provider<TestModule>>::Interface> for TestModule {}
```

for whatever reason `dyn Repository` breaks, `()` does not.

why this compiles in the old solver:
- before equating impl headers, normalize both impls
- `Projection(<ServiceImpl as Provider<TestModule>>::Interface)`
- single candidate with nested goal `TestModule: HasProvider<&'static ()>` is simply emitted as a nested goal, successfully normalize to `Service`
- equating `impl HasProvider<&'static ()> for TestModule {}` with `impl HasProvider<Service> for TestModule {}` fails

what's happening in the new solver:
- coherence equates the two impls, resulting in a nested `alias-relate(<ServiceImpl as Provider<TestModule>>::Interface, &'static ())`
- uniquification results in `exists<'0> alias-relate(<ServiceImpl as Provider<TestModule>>::Interface, &'0 ())`
- normalizes-to candidat with single impl, normalizing to `Service` with nested goal `TestModule: HasProvider<&'static ()>`, again, canonicalized to `'0`
- two candidates:
    - `impl HasProvider<&'static ()> for TestModule {}`, YES `'0 == 'static`
    - `impl HasProvider<<ServiceImpl as Provider<TestModule>>::Interface> for TestModule {}`
        - unify `<ServiceImpl as Provider<TestModule>>::Interface` with `&'static ()`, resulting in an inductive cycle
    - AMBIG
- WHY DOES ALIAS-RELATE NOT FAIL?

### PR DESCRIPTION
**Pull request by lcnr**
Do not erase overflow constraints if they are from equating the impl header when normalizing[^1].

This should be the minimal change to not break crates depending on the old project behavior of "apply impl constraints while only lazily evaluating any nested goals".

Fixes https://github.com/rust-lang/trait-system-refactor-initiative/issues/70, see https://hackmd.io/ATf4hN0NRY-w2LIVgeFsVg for the reasoning behind this.

Only keeping constraints on overflow for `normalize-to` goals as that's the only thing needed for backcompat. It also allows us to not track the origin of root obligations. The issue with root goals would be something like the following:

```rust
trait Foo {}
trait Bar {}

trait FooBar {}
impl<T: Foo + Bar> FooBar for T {}

// These two should behave the same, rn we can drop constraints for both,
// but if we don't drop `Misc` goals we would only drop the constraints for
// `FooBar` unless we track origins of root obligations.
fn func1<T: Foo + Bar>() {}
fn func2<T: FooBaz>() {}
```

[^1]: mostly, the actual rules are slightly different

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs [hunk 1]
```diff
@@ -386,6 +388,30 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
...
     }
 
+    fn instantiate_response_discarding_overflow(
+        &mut self,
+        param_env: ty::ParamEnv<'tcx>,
+        source: GoalSource,
+        original_values: Vec<ty::GenericArg<'tcx>>,
+        response: CanonicalResponse<'tcx>,
+    ) -> Result<(Certainty, bool, Vec<Goal<'tcx, ty::Predicate<'tcx>>>), NoSolution> {
+        let keep_overflow_constraints = || {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs:398]: needs comment:

doing this only for normalize-to goals as that's the only thing needed for backcompat

doesn't require us to track the origin of a root obligation. If we were to also do this for trait goals, we'd have to worry about

```rust
trait Foo {}
trait Bar {}

trait FooBar {}
impl<T: Foo + Bar> FooBar for T {}

// These two should behave the same, rn we can drop constraints for both,
// but if we don't drop `Misc` goals we would only drop the constraints for
// `FooBar`
fn func1<T: Foo + Bar>() {}
fn func2<T: FooBaz>() {}
```

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs:398]: pls add comment

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me after adding that comment you said you would

**Pull request merged**